### PR TITLE
refactor(api): remove rx from `boot/challenge.js`

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

A basic start

### CODE REVIEW

File: api-server/src/server/boot/challenge.js [hunk 1]
```diff
@@ -273,63 +272,55 @@ export function isValidChallengeCompletion(req, res, next) {
   return next();
 }
 
-export function modernChallengeCompleted(req, res, next) {
+export async function modernChallengeCompleted(req, res) {
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:275]: Express doesn't understand async handlers.  It doesn't `await` them, so error handling won't work properly.

We can either use a library to wrap async handlers or just write our own given how simple it is.

See https://expressjs.com/en/guide/error-handling.html for reference.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api-server/src/server/boot/challenge.js (hunk 1 above)
Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:275]: Ah, thank you for that. Now it makes sense why rx used the `next` cb.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:275]: No worries.

To be honest, I was a bit mystified by that, but I guess the second argument must be the error handler.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:275]: Also, the same argument applies to https://github.com/freeCodeCamp/freeCodeCamp/pull/49386, but I didn't want to derail the discussion there.

File: api-server/src/server/boot/challenge.js [hunk 2]
```diff
@@ -273,63 +272,60 @@ export function isValidChallengeCompletion(req, res, next) {
...
-        multifileCertProjectIds.includes(id)
-      ) {
-        completedChallenge.challengeType = challengeType;
-      }
+  try {
+    // This is an ugly way to update `user.completedChallenges`
+    await user.getCompletedChallenges$().toPromise();
+  } catch (e) {
+    return next(e);
+  }
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:283]: What does this actually do? I couldn't see anything obviously broken when I commented out this section.

File: api-server/src/server/boot/challenge.js [hunk 3]
```diff
@@ -273,63 +272,60 @@ export function isValidChallengeCompletion(req, res, next) {
...
+
+  const { alreadyCompleted, savedChallenges, updateData } = buildUserUpdate(
+    user,
+    id,
+    completedChallenge
+  );
+
+  const points = alreadyCompleted ? user.points : user.points + 1;
+
+  user.updateAttributes(updateData, (err, _updatedUser) => {
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:314]:
```suggestion
  user.updateAttributes(updateData, err => {
```
Seems redundant if we're not doing anything with the updated user.

File: api-server/src/server/boot/challenge.js [hunk 4]
```diff
@@ -370,69 +366,64 @@ function projectCompleted(req, res, next) {
...
-        user,
-        completedChallenge.id,
-        completedChallenge
-      );
+  try {
+    // This is an ugly hack to update `user.completedChallenges`
+    await user.getCompletedChallenges$().toPromise();
+  } catch (e) {
+    return next(e);
+  }
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:374]: Same point as above - is this necessary?

File: api-server/src/server/boot/challenge.js [hunk 5]
```diff
@@ -370,69 +366,64 @@ function projectCompleted(req, res, next) {
...
-      });
-    })
-    .subscribe(() => {}, next);
+  const { alreadyCompleted, updateData } = buildUserUpdate(
+    user,
+    completedChallenge.id,
+    completedChallenge
+  );
+
+  user.updateAttributes(updateData, (err, _updatedUser) => {
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:382]:
```suggestion
  user.updateAttributes(updateData, err => {
```

File: api-server/src/server/boot/challenge.js [hunk 6]
```diff
@@ -370,69 +366,64 @@ function projectCompleted(req, res, next) {
...
-      const { alreadyCompleted, updateData } = buildUserUpdate(
-        user,
-        completedChallenge.id,
-        completedChallenge
-      );
+  try {
+    await user.getCompletedChallenges$().toPromise();
+  } catch (e) {
+    return next(e);
+  }
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:405]: And again.

File: api-server/src/server/boot/challenge.js [hunk 7]
```diff
@@ -449,42 +440,37 @@ function saveChallenge(req, res, next) {
...
+  } else {
+    $push.savedChallenges = challengeToSave;
+    savedChallenges.push(challengeToSave);
+  }
+
+  const updateData = {};
+  if (!isEmpty($set)) updateData.$set = $set;
+  if (!isEmpty($push)) updateData.$push = $push;
+
+  user.updateAttributes(updateData, (err, _updatedUser) => {
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:465]:
```suggestion
  user.updateAttributes(updateData, err => {
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api-server/src/server/boot/challenge.js (hunk 3 above)
Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:314]: True, but I just like knowing it is there. Still want me to remove it?

File: api-server/src/server/boot/challenge.js (hunk 4 above)
Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:374]: Yes

File: api-server/src/server/boot/challenge.js [hunk 8]
```diff
@@ -370,69 +366,64 @@ function projectCompleted(req, res, next) {
...
+  );
+
+  user.updateAttributes(updateData, (err, updatedUser) => {
+    if (err) {
+      return next(err);
+    }
+
+    return res.json({
+      alreadyCompleted,
+      points: alreadyCompleted ? updatedUser.points : updatedUser.points + 1,
```
[the remark this answers is not in the document]
Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:420]: 🤷‍♂️  At least this knows if the update is successful, and does not return an updated `points` when the update fails?

This is as it is before my change - beyond the scope of this PR (removing RX)

File: api-server/src/server/boot/challenge.js (hunk 2 above)
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:283]: I know it runs that code, I just don't understand what purpose it serves.

To put it another way: under what circumstances is `user.completedChallenges` in need of an update?

File: api-server/src/server/boot/challenge.js (hunk 3 above)
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:314]: I don't mind it that much.

File: api-server/src/server/boot/challenge.js (hunk 4 above)
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:374]: :thinking:

File: api-server/src/server/boot/challenge.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:420]: > This is as it is before my change

I don't think that's the case.

https://github.com/freeCodeCamp/freeCodeCamp/blob/080ee3539083203fb313f231d2ee3589ad771756/api-server/src/server/boot/challenge.js#L402-L432

The user is `req.user` at all points, same as for the other handlers.

File: api-server/src/server/boot/challenge.js (hunk 2 above)
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:283]: It's a little out of scope, so if you'd rather we address it later, that's fine with me.

Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:283]: If you are asking why/how `completedChallenges` got updated between the initial fetch of the `user` object and this line of code, then I do not know:

image

Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:283]: image

Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:283]: Fair. Let's handle that in another PR.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM 👍

One bit of rx removal at a time...

**Pull request merged**